subprojects/docs/src/docs/userguide/authoring-builds/more_about_tasks.adoc

```diff
@@ -830,20 +819,41 @@ Let’s say we don’t want to run the `processTemplates` task if there are no s
 
 .Using skipWhenEmpty() via the runtime API
 ====
-include::sample[dir="userguide/tasks/incrementalBuild/customTaskClass/groovy",files="build.gradle[tags=runtime-api-conf]"]
-include::sample[dir="userguide/tasks/incrementalBuild/customTaskClass/kotlin",files="build.gradle.kts[tags=runtime-api-conf]"]
+include::sample[dir="userguide/tasks/incrementalBuild/customTaskClass/groovy",files="build.gradle[tags=ad-hoc-task-skip-when-empty]"]
+include::sample[dir="userguide/tasks/incrementalBuild/customTaskClass/kotlin",files="build.gradle.kts[tags=ad-hoc-task-skip-when-empty]"]
 
-.Output of **`gradle clean processTemplatesRuntimeConf`**
+.Output of **`gradle clean processTemplatesAdHocSkipWhenEmpty`**
 ----
-> gradle clean processTemplatesRuntimeConf
-include::{samplesPath}/userguide/tasks/incrementalBuild/customTaskClass/runtimeIncrementalApiConfiguration.out[]
+> gradle clean processTemplatesAdHocSkipWhenEmpty
+include::{samplesPath}/userguide/tasks/incrementalBuild/customTaskClass/incrementalAdHocTaskNoSource.out[]
 ----
 ====
 
 The `TaskInputs.files()` method returns a builder that has a `skipWhenEmpty()` method. Invoking this method is equivalent to annotating to the property with <<#skip-when-empty,`@SkipWhenEmpty`>>.
 
 Now that you have seen both the annotations and the runtime API, you may be wondering which API you should be using. Our recommendation is to use the annotations wherever possible, and it’s sometimes worth creating a custom task class just so that you can make use of them. The runtime API is more for situations in which you can’t use the annotations.
 
+[[sec:runtime_api_for_custom_tasks]]
+===== Using it for custom task types
+
+Another type of example involves registering additional inputs and outputs for instances of a custom task class.
+For example, imagine that the `ProcessTemplates` task also needs to read `src/headers/headers.txt` (e.g. because it is included from one of the sources).
+You'd want Gradle to know about this input file, so that it can re-execute the task whenever the contents of this file change.
+With the runtime API you can do just that:
+
+.Using runtime API with custom task type
+====
+include::sample[dir="userguide/tasks/incrementalBuild/customTaskClass/groovy",files="build.gradle[tags=custom-class-runtime-api]"]
+include::sample[dir="userguide/tasks/incrementalBuild/customTaskClass/kotlin",files="build.gradle.kts[tags=custom-class-runtime-api]"]
+====
+
+Using the runtime API like this is a little like using `doLast()` and `doFirst()` to attach extra actions to a task, except in this case we’re attaching information about inputs and outputs.
+
+[WARNING]
+====
+If the task type is already using the incremental build annotations, registering inputs or outputs with the same property names will result in an error.
```

Comment:
You cannot override file inputs and outputs, though you can override non-file inputs by inputs registered via the runtime API. I don't know why this is the case, though it currently works like this.

Reply:
We should deprecate overriding non-file properties, too. I'll create an issue.

Reply:
https://github.com/gradle/gradle/issues/10963

Reply:
I'd leave this as it is. WDYT?

Reply:
Fine with me.